Technological Institute of the Philippines | Quezon City - Computer Engineering
--- | ---
Course Code: | CPE 019
Code Title: | Emerging Technologies 2 in CpE
1st Semester | AY 2023-2024
<hr> | <hr>
<u>**Assignment 7.1 : Classifications and Regression** |
**Name**: | Niemo, Christian Al C.
**Section**: | CPE32S9
**Date Performed**: | 04-14-24
**Date Submitted**: | 04-15-24
**Instructor**: | Engr. Roman Richard
<hr> | <hr>

**Installing all necessary packages**

In [ ]:
!pip install scikeras

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 17.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.1/12.1 MB 58.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 311.2/311.2 kB 25.3 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.2.2
    Uninstalling scikit-learn-1.2.2:
      Successfully uninstalled scikit-learn-1.2.2
  Attempting uninstall: keras
    Found existing installation: keras 2.15.0
    Uninstalling keras-2.15.0:
      Successfully uninstalled keras-2.15.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.15.0 requires keras<2.16,>=2.15.0, but you have keras 3.2.1 which is incompatible.


In [ ]:
!pip install np_utils

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.0/62.0 kB 1.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for np_utils: filename=np_utils-0.6.0-py3-none-any.whl size=56441 sha256=8cf91bd663fe5a6ddad64cc454ab356f021efbec84f5b0f8faeffecebd256125
  Stored in directory: /root/.cache/pip/wheels/b6/c7/50/2307607f44366dd021209f660045f8d51cb976514d30be7cc7
Successfully built np_utils


**Mounting the drive to colab**

In [ ]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


**Importing the libraries**

In [ ]:
import numpy as np
import pandas as pd
from keras.models import Sequential
from keras.layers import Dense
from scikeras.wrappers import KerasClassifier
from scikeras.wrappers import KerasRegressor
from keras.utils import to_categorical
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

***Classification Dataset***

Obesity Levels - obesity is a significant global health problem that has both physical and mental consequences. Its prevalence is steadily increasing, necessitating new research into the factors influencing obesity and how to predict its occurrence based on those factors.

In [ ]:
#reading the csv file
df = pd.read_csv('/content/drive/My Drive/Colab Notebooks/class.csv')
df.head()

Age  Gender  Height  Weight        CALC FAVC  FCVC  NCP  SCC SMOKE  CH2O  \
0  21.0  Female    1.62    64.0          no   no   2.0  3.0   no    no   2.0   
1  21.0  Female    1.52    56.0   Sometimes   no   3.0  3.0  yes   yes   3.0   
2  23.0    Male    1.80    77.0  Frequently   no   2.0  3.0   no    no   2.0   
3  27.0    Male    1.80    87.0  Frequently   no   3.0  3.0   no    no   2.0   
4  22.0    Male    1.78    89.8   Sometimes   no   2.0  1.0   no    no   2.0   

  family_history_with_overweight  FAF  TUE       CAEC                 MTRANS  \
0                            yes  0.0  1.0  Sometimes  Public_Transportation   
1                            yes  3.0  0.0  Sometimes  Public_Transportation   
2                            yes  2.0  1.0  Sometimes  Public_Transportation   
3                             no  2.0  0.0  Sometimes                Walking   
4                             no  0.0  0.0  Sometimes  Public_Transportation   

            NObeyesdad  
0        Normal_Weight  
1        Normal_Weight  
2        Normal_Weight  
3   Overweight_Level_I  
4  Overweight_Level_II

In [ ]:
#changing the string values into numerical values using lambda
df["Gender"] = df["Gender"].apply(lambda toLabel:0 if toLabel == 'Male' else 1)
df["FAVC"] = df["FAVC"].apply(lambda toLabel:0 if toLabel == 'no' else 1)
df["SCC"] = df["SCC"].apply(lambda toLabel:0 if toLabel == 'no' else 1)
df["SMOKE"] = df["SMOKE"].apply(lambda toLabel:0 if toLabel == 'no' else 1)
df["family_history_with_overweight"] = df["family_history_with_overweight"].apply(lambda toLabel:0 if toLabel == 'no' else 1)
df.head()

Age  Gender  Height  Weight        CALC  FAVC  FCVC  NCP  SCC  SMOKE  \
0  21.0       1    1.62    64.0          no     0   2.0  3.0    0      0   
1  21.0       1    1.52    56.0   Sometimes     0   3.0  3.0    1      1   
2  23.0       0    1.80    77.0  Frequently     0   2.0  3.0    0      0   
3  27.0       0    1.80    87.0  Frequently     0   3.0  3.0    0      0   
4  22.0       0    1.78    89.8   Sometimes     0   2.0  1.0    0      0   

   CH2O  family_history_with_overweight  FAF  TUE       CAEC  \
0   2.0                               1  0.0  1.0  Sometimes   
1   3.0                               1  3.0  0.0  Sometimes   
2   2.0                               1  2.0  1.0  Sometimes   
3   2.0                               0  2.0  0.0  Sometimes   
4   2.0                               0  0.0  0.0  Sometimes   

                  MTRANS           NObeyesdad  
0  Public_Transportation        Normal_Weight  
1  Public_Transportation        Normal_Weight  
2  Public_Transportation        Normal_Weight  
3                Walking   Overweight_Level_I  
4  Public_Transportation  Overweight_Level_II

In [ ]:
#changing the string values into categorical types
df['CALC'] = df['CALC'].astype('category')
df['CALC'] = df['CALC'].cat.codes

df['CAEC'] = df['CAEC'].astype('category')
df['CAEC'] = df['CAEC'].cat.codes

df['MTRANS'] = df['MTRANS'].astype('category')
df['MTRANS'] = df['MTRANS'].cat.codes

df.head()

Age  Gender  Height  Weight  CALC  FAVC  FCVC  NCP  SCC  SMOKE  CH2O  \
0  21.0       1    1.62    64.0     3     0   2.0  3.0    0      0   2.0   
1  21.0       1    1.52    56.0     2     0   3.0  3.0    1      1   3.0   
2  23.0       0    1.80    77.0     1     0   2.0  3.0    0      0   2.0   
3  27.0       0    1.80    87.0     1     0   3.0  3.0    0      0   2.0   
4  22.0       0    1.78    89.8     2     0   2.0  1.0    0      0   2.0   

   family_history_with_overweight  FAF  TUE  CAEC  MTRANS           NObeyesdad  
0                               1  0.0  1.0     2       3        Normal_Weight  
1                               1  3.0  0.0     2       3        Normal_Weight  
2                               1  2.0  1.0     2       3        Normal_Weight  
3                               0  2.0  0.0     2       4   Overweight_Level_I  
4                               0  0.0  0.0     2       3  Overweight_Level_II

In [ ]:
#splitting the input values into output values
#X = all input values
#Y = output value
dataset = df.values
X = dataset[:,0:16]
Y = dataset[:,16]

In [ ]:
#printing the values in array form
X

array([[21.0, 1, 1.62, ..., 1.0, 2, 3],
       [21.0, 1, 1.52, ..., 0.0, 2, 3],
       [23.0, 0, 1.8, ..., 1.0, 2, 3],
       ...,
       [22.524036, 1, 1.752206, ..., 0.646288, 2, 3],
       [24.361936, 1, 1.73945, ..., 0.586035, 2, 3],
       [23.664709, 1, 1.738836, ..., 0.714137, 2, 3]], dtype=object)

In [ ]:
#printing all the output values per row
Y

array(['Normal_Weight', 'Normal_Weight', 'Normal_Weight', ...,
       'Obesity_Type_III', 'Obesity_Type_III', 'Obesity_Type_III'],
      dtype=object)

In [ ]:
#one-hot coded variables
#encode class values as integers
encoder = LabelEncoder()
encoder.fit(Y)
encoded_Y = encoder.transform(Y)
#convert integers to dummy variables
dummy_y = to_categorical(encoded_Y)

In [ ]:
# define baseline model
def baseline_model():
# create model
  model = Sequential()
  model.add(Dense(4, input_shape=(16,), activation='relu'))
  model.add(Dense(7, activation='sigmoid'))
# Compile model
  model.compile(loss='categorical_crossentropy' , optimizer= 'adam' , metrics=[ 'accuracy' ])
  return model

In [ ]:
#training the model
estimator = KerasClassifier(build_fn=baseline_model, epochs=200, batch_size=5, verbose=0)
#evaluating the model using k-fold cross validation multiclass classification
kfold = KFold(n_splits=10, shuffle=True)
results = cross_val_score(estimator, X, dummy_y, cv=kfold)
print("Baseline: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using 

Baseline: 65.37% (24.79%)


A baseline neural network model is defined using the Keras library. The model consists of one hidden layer with 4 neurons, an input shape of 16, and an activation function of ReLU. The output layer has 7 neurons and uses the sigmoid activation function. The purpose of this model is to perform multi-class classification.

To train and evaluate the model, the code utilizes k-cross validation with 10 folds. The model is trained for 200 epochs with a batch size of 5, using the Adam optimizer and categorical cross-entropy loss function. The accuracy metric is used to assess the model's performance.

The results of the cross-validation are displayed, showing the mean accuracy obtained across the 10 folds as a percentage. The standard deviation is also provided, indicating the variance in performance across the folds.

Based on the results obtained from this baseline model, it can be concluded that the model achieved a low level of accuracy in classifying the given data. Further improvements or modifications to the model architecture or hyperparameters may be necessary to enhance its performance.

In [ ]:
# define baseline model
def baseline_model():
# create model
  model = Sequential()
  model.add(Dense(18, input_shape=(16,), activation='relu'))
  model.add(Dense(9, activation='relu'))
  model.add(Dense(7, activation='sigmoid'))
# Compile model
  model.compile(loss='categorical_crossentropy' , optimizer= 'adam' , metrics=[ 'accuracy' ])
  return model

#training the model
estimator1 = KerasClassifier(build_fn=baseline_model, epochs=200, batch_size=5, verbose=0)
#evaluating the model using k-fold cross validation multiclass classification
kfold1 = KFold(n_splits=10, shuffle=True)
results1 = cross_val_score(estimator1, X, dummy_y, cv=kfold1)
print("Baseline: %.2f%% (%.2f%%)" % (results1.mean()*100, results1.std()*100))

/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using 

Baseline: 68.79% (9.19%)


In this part, I defined a updated baseline neural network model. The model consists of one hidden layer with 18 neurons, an input shape of 16, and an activation function of ReLU. The output layer has 7 neurons and uses the sigmoid activation function.

Considering the baseline result obtained, which is a little bit low, further modifications were made to improve the network's performance. One approach was to make the network deeper and larger. By increasing the number of layers and neurons, the model may be able to capture more complex patterns and improve its ability to classify the weight categories accurately.

In [ ]:
# define baseline model
def baseline_model():
# create model
  model = Sequential()
  model.add(Dense(18, input_shape=(16,), activation='relu'))
  model.add(Dense(9, activation='relu'))
  model.add(Dense(5, activation='relu'))
  model.add(Dense(7, activation='softmax'))
# Compile model
  model.compile(loss='categorical_crossentropy' , optimizer= 'adam' , metrics=[ 'accuracy' ])
  return model

#training the model
estimator = KerasClassifier(build_fn=baseline_model, epochs=200, batch_size=5, verbose=0)
#evaluating the model using k-fold cross validation in multiclass classification
kfold = KFold(n_splits=10, shuffle=True)
results = cross_val_score(estimator, X, dummy_y, cv=kfold)
print("Baseline: %.2f%% (%.2f%%)" % (results.mean()*100, results.std()*100))

/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/scikeras/wrappers.py:925: UserWarning: ``build_fn`` will be renamed to ``model`` in a future release, at which point use of ``build_fn`` will raise an Error instead.
  X, y = self._initialize(X, y)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using 

Baseline: 74.13% (27.98%)


In this code, a revised baseline neural network model is defined. The model architecture has been modified to include additional layers and neurons. The revised model consists of:

    One hidden layer with 18 neurons, using the ReLU activation function.
    Another hidden layer with 9 neurons, also using the ReLU activation function.
    An additional hidden layer with 5 neurons, using the ReLU activation function.
    The output layer with 7 neurons, using the softmax activation function.

The softmax activation function is commonly used for multi-class classification tasks as it produces a probability distribution over the output classes.

Based on the results obtained from this revised baseline model, it can be concluded that the model achieved a certain level of accuracy in classifying the given data. 74% is a huge improvement, coming from the 65% from the first version of the code.

***Regression Dataset***

Medical Insurance Cost Prediction - the medical insurance dataset contains a wide range of factors that influence the cost of medical expenses, including age, sex, BMI, smoking status, number of children, and region. The dataset is designed to be used for training machine learning models that can predict medical expenses for new policyholders.

In [ ]:
#reading the csv file
df = pd.read_csv('/content/drive/My Drive/Colab Notebooks/reg.csv')
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [ ]:
#changing the string values into numerical values using lambda
df["sex"] = df["sex"].apply(lambda toLabel:0 if toLabel == 'male' else 1)
df["smoker"] = df["smoker"].apply(lambda toLabel:0 if toLabel == 'no' else 1)
df["region"] = df["region"].apply(lambda toLabel:0 if toLabel == 'northwest' else 1)
df.head()

age  sex     bmi  children  smoker  region      charges
0   19    1  27.900         0       1       1  16884.92400
1   18    0  33.770         1       0       1   1725.55230
2   28    0  33.000         3       0       1   4449.46200
3   33    0  22.705         0       0       0  21984.47061
4   32    0  28.880         0       0       0   3866.85520

In [ ]:
#splitting the input values into output values
#X = all input values
#Y = output value
dataset = df.values
X = dataset[:,0:6]
Y = dataset[:,6]

In [ ]:
#printing the values in array form
X

array([[19.  ,  1.  , 27.9 ,  0.  ,  1.  ,  1.  ],
       [18.  ,  0.  , 33.77,  1.  ,  0.  ,  1.  ],
       [28.  ,  0.  , 33.  ,  3.  ,  0.  ,  1.  ],
       ...,
       [18.  ,  1.  , 36.85,  0.  ,  0.  ,  1.  ],
       [21.  ,  1.  , 25.8 ,  0.  ,  0.  ,  1.  ],
       [61.  ,  1.  , 29.07,  0.  ,  1.  ,  0.  ]])

In [ ]:
#printing all the output values per row
Y

array([16884.924 ,  1725.5523,  4449.462 , ...,  1629.8335,  2007.945 ,
       29141.3603])

In [ ]:
# define base model
def baseline_model():
	# create model
  model = Sequential()
  model.add(Dense(6, input_shape=(6,), kernel_initializer='normal', activation='relu'))
  model.add(Dense(1, kernel_initializer='normal'))
	# Compile model
  model.compile(loss='mean_squared_error', optimizer='adam')
  return model

#training the model
estimator = KerasRegressor(model=baseline_model, epochs=100, batch_size=5, verbose=0)
#evaluating the model using k-fold cross validation in regression
kfold = KFold(n_splits=10)
results = cross_val_score(estimator, X, Y, cv=kfold, scoring='neg_mean_squared_error')
print("Baseline: %.2f (%.2f) MSE" % (results.mean(), results.std()))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/lo

Baseline: -128091886.28 (15188026.40) MSE


A baseline neural network model for regression tasks is defined using the Keras library. The model architecture is designed to predict continuous numerical values.

The baseline model consists of:

    One hidden layer with 6 neurons, using the ReLU activation function. The kernel_initializer parameter is set to 'normal', which initializes the weights of the layer using a normal distribution.
    The output layer with 1 neuron, representing the predicted numerical value. The kernel_initializer parameter is set to 'normal' as well.

The code implements a baseline model for regression tasks, utilizing k-fold cross-validation with 10 folds. The model is trained for 100 epochs with a batch size of 5, using the mean squared error (MSE) loss function and the Adam optimizer. The evaluation metric used is also the MSE. Across the 10 folds, the baseline model achieved a negative MSE of approximately -128091886.28, with a standard deviation of 15188026.40.

In [ ]:
# define base model
def baseline_model():
	# create model
  model = Sequential()
  model.add(Dense(6, input_shape=(6,), kernel_initializer='normal', activation='relu'))
  model.add(Dense(1, kernel_initializer='normal'))
	# Compile model
  model.compile(loss='mean_squared_error', optimizer='adam')
  return model

#standardizing the model
estimators = []
estimators.append(('standardize', StandardScaler()))
#training the model
estimators.append(('mlp', KerasRegressor(model=baseline_model, epochs=50, batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
#evaluating the model using k-fold cross validation in regression
kfold = KFold(n_splits=10)
results = cross_val_score(pipeline, X, Y, cv=kfold, scoring='neg_mean_squared_error')
print("Standardized: %.2f (%.2f) MSE" % (results.mean(), results.std()))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/lo

Standardized: -285235980.84 (29584315.44) MSE


The input data is standardized before training the model to potentially improve its performance.

Here are the modifications to the code:

    A StandardScaler is added as a preprocessing step in the pipeline. This scaler standardizes the input data, which can be beneficial for neural networks, especially when the features have different scales.

    The Pipeline class from scikit-learn is used to chain the preprocessing step (standardization) with the neural network model.

    The model is trained and evaluated using k-fold cross-validation, similar to the previous example. The pipeline object is passed to the cross_val_score function.

Based on the revised code, the baseline model with standardized input yielded a mean squared error (MSE) of approximately -285235980.84, with a standard deviation of 29584315.44 across the 10 folds of the cross-validation.

The updated model, featuring standardized input data, exhibits a reduced MSE in comparison to the previous baseline model. This outcome indicates that the process of standardizing the input data may have enhanced the model's ability to accurately predict numerical values.

In [ ]:
#define base model (deeper network)
def baseline_model():
	#create model
  model = Sequential()
  model.add(Dense(6, input_shape=(6,), kernel_initializer='normal', activation='relu'))
  model.add(Dense(3, kernel_initializer='normal', activation='relu'))
  model.add(Dense(1, kernel_initializer='normal'))
	#Compile model
  model.compile(loss='mean_squared_error', optimizer='adam')
  return model

#standardizing the model
estimators = []
estimators.append(('standardize', StandardScaler()))
#training the model
estimators.append(('mlp', KerasRegressor(model=baseline_model, epochs=50, batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
#evaluating the model using k-fold cross validation in regression
kfold = KFold(n_splits=10)
results = cross_val_score(pipeline, X, Y, cv=kfold, scoring='neg_mean_squared_error')
print("Standardized: %.2f (%.2f) MSE" % (results.mean(), results.std()))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/lo

Standardized: -67079174.57 (67958390.74) MSE


The baseline neural network model for regression tasks has been further modified. Here are the key changes:

    The model architecture now includes an additional hidden layer. The revised model consists of:

      One hidden layer with 6 neurons, using the ReLU activation function.
      Another hidden layer with 3 neurons, also using the ReLU activation function.
      The output layer with 1 neuron, representing the predicted numerical value.

Based on the obtained results, the baseline model with standardized input achieved a mean squared error (MSE) of approximately -67079174.57, with a standard deviation of 67958390.74 across the 10 folds of the cross-validation.

Comparing this to the previous model, it is observed that the addition of an extra hidden layer with 3 neurons has led to a different level of performance. Specifically, the MSE has decreased, indicating an improvement in the model's ability to predict numerical values. This suggests that the modification to the model architecture has positively impacted its performance in the given regression task.

In [ ]:
#define base model (smaller neurons)
def baseline_model():
	# create model
  model = Sequential()
  model.add(Dense(3, input_shape=(6,), kernel_initializer='normal', activation='relu'))
  model.add(Dense(1, kernel_initializer='normal'))
	# Compile model
  model.compile(loss='mean_squared_error', optimizer='adam')
  return model

#standardizing the model
estimators = []
estimators.append(('standardize', StandardScaler()))
#training the model
estimators.append(('mlp', KerasRegressor(model=baseline_model, epochs=50, batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
#evaluating the model using k-fold cross validation in regression
kfold = KFold(n_splits=10)
results = cross_val_score(pipeline, X, Y, cv=kfold, scoring='neg_mean_squared_error')
print("Standardized: %.2f (%.2f) MSE" % (results.mean(), results.std()))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/lo

Standardized: -304616561.67 (31012246.77) MSE


In this updated baseline neural network model, regression tasks has been further modified. Here are the key changes:

    The model architecture has been simplified. The revised model consists of:
        One hidden layer with 3 neurons, using the ReLU activation function, which is smaller from example before.
        The output layer with 1 neuron, representing the predicted numerical value.

The revised baseline model, which incorporates standardized input, performed well in terms of mean squared error (MSE) during the k-fold cross-validation. The MSE was observed to be around -304616561.67, with a standard deviation of 31012246.77 across the 10 folds. These results indicate that the modifications made to the code, including the standardization of input, have had a positive impact on the model's performance.

In [ ]:
# define base model (larger neurons)
def baseline_model():
	# create model
  model = Sequential()
  model.add(Dense(12, input_shape=(6,), kernel_initializer='normal', activation='relu'))
  model.add(Dense(1, kernel_initializer='normal'))
	# Compile model
  model.compile(loss='mean_squared_error', optimizer='adam')
  return model

#standardizing the model
estimators = []
estimators.append(('standardize', StandardScaler()))
#training the model
estimators.append(('mlp', KerasRegressor(model=baseline_model, epochs=50, batch_size=5, verbose=0)))
pipeline = Pipeline(estimators)
#evaluating the model using k-fold cross validation in regression
kfold = KFold(n_splits=10)
results = cross_val_score(pipeline, X, Y, cv=kfold, scoring='neg_mean_squared_error')
print("Standardized: %.2f (%.2f) MSE" % (results.mean(), results.std()))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:86: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/lo

Standardized: -252787119.50 (27587943.89) MSE


The code provided implements a baseline neural network model and evaluates its performance using k-fold cross-validation in a regression task. The model architecture consists of an input layer with 6 units, a hidden layer with 12 units using the ReLU activation function, and an output layer with 1 unit.

The results of the k-fold cross-validation indicate that the average (mean) negative mean squared error achieved by the model is -252787119.50, with a standard deviation of 27587943.89. The negative mean squared error is used because of how the scoring metric is defined in scikit-learn.

Based on these results, it can be concluded that the baseline neural network model, when trained and evaluated using k-fold cross-validation, achieves a relatively low mean squared error on the regression task.

## **Conclusion**

In order to create a classification model, a base model must first be created and then its performance must be assessed using k-fold cross-validation. More hidden layers can be added to the model to increase its accuracy and strengthen its capacity to identify complex patterns and correlations in the data. These procedures can be used to create a classification model that is more accurate in predicting and classifying data according to the specified qualities.


The following procedures should be followed when performing regression analysis: first, establish a baseline model as a point of reference; second, improve the model's performance by standardizing the dataset to guarantee that each feature contributes equally; and third, investigate layer and neuron tuning by assessing both smaller and larger network configurations. By improving the architecture and data prior treatment methods, these phases seek to increase the regression model's efficacy and accuracy.